src/hooks/routing/useRouterTrade.ts

```diff
@@ -0,0 +1,141 @@
+// eslint-disable-next-line no-restricted-imports
+import type { ChainId } from '@uniswap/smart-order-router'
+import { useRouterArguments } from 'hooks/routing/useRouterArguments'
+import useActiveWeb3React from 'hooks/useActiveWeb3React'
+import useDebounce from 'hooks/useDebounce'
+import useIsValidBlock from 'hooks/useIsValidBlock'
+import useIsWindowVisible from 'hooks/useIsWindowVisible'
+import { useStablecoinAmountFromFiatValue } from 'hooks/useUSDCPrice'
+import ms from 'ms.macro'
+import { useMemo } from 'react'
+import { useGetQuoteQuery } from 'state/routing/slice'
+import { GetQuoteResult, InterfaceTrade, TradeState } from 'state/routing/types'
+import { computeRoutes, transformRoutesToTrade } from 'state/routing/utils'
+
+import { isAutoRouterSupportedChain } from './clientSideSmartOrderRouter'
+
+export const INVALID_TRADE = { state: TradeState.INVALID, trade: undefined }
+
+/**
+ * Returns the best trade by invoking the routing api or the smart order router on the client
+ * @param tradeType whether the swap is an exact in/out
+ * @param routerUrl the base URL of the integrator's auto router API
+ * @param amountSpecified the exact amount to swap in/out
+ * @param otherCurrency the desired output/payment currency
+ */
+export function useRouterTrade<TTradeType extends TradeType>(
+  tradeType: TTradeType,
+  routerUrl?: string,
+  amountSpecified?: CurrencyAmount<Currency>,
+  otherCurrency?: Currency
+): {
+  state: TradeState
+  trade: InterfaceTrade<Currency, Currency, TTradeType> | undefined
+} {
+  const { chainId, library } = useActiveWeb3React()
+  const autoRouterSupported = isAutoRouterSupportedChain(chainId)
+  const isWindowVisible = useIsWindowVisible()
+  // Debounce is used to prevent excessive requests to SOR, as it is data intensive.
+  // Fast user actions (ie updating the input) should be debounced, but currency changes should not.
+  const [debouncedAmount, debouncedOtherCurrency] = useDebounce(
+    useMemo(() => [amountSpecified, otherCurrency], [amountSpecified, otherCurrency]),
+    200
+  )
+  const debouncedAmountSpecified = autoRouterSupported && isWindowVisible ? debouncedAmount : undefined
+
+  const [currencyIn, currencyOut]: [Currency | undefined, Currency | undefined] = useMemo(
+    () =>
+      tradeType === TradeType.EXACT_INPUT
+        ? [debouncedAmountSpecified?.currency, debouncedOtherCurrency]
+        : [debouncedOtherCurrency, debouncedAmountSpecified?.currency],
+    [debouncedAmountSpecified, debouncedOtherCurrency, tradeType]
+  )
+
+  const currencyChainId = currencyIn?.chainId as ChainId
+  if (currencyChainId && !isAutoRouterSupportedChain(currencyChainId)) {
+    throw new Error(`Router does not support this token's chain (chainId: ${currencyChainId}).`)
+  }
+
+  const queryArgs = useRouterArguments({
+    tokenIn: currencyIn,
+    tokenOut: currencyOut,
+    amount: debouncedAmountSpecified,
+    tradeType,
+    routerUrl,
+    provider: library as JsonRpcProvider,
```

Comment:
should we somehow verify that `library` is in fact a `jsonrpcprovider` instead of casting?

Reply:
and throw error if not

Reply:
library is of type `(JsonRpcProvider & { provider?: ExternalProvider }) | Web3Provider` and Web3Provider extends JsonRpcProvider

Reply:
so we should be good